ecs/src/entities.zig

```diff
@@ -327,361 +327,397 @@ pub const void_archetype_hash = std.math.maxInt(u64);
+            if (archetype.len > 1) {
+                const last_row_entity_id = archetype.get(entities.allocator, archetype.len - 1, ".id", EntityID).?;
+                try entities.entities.put(entities.allocator, last_row_entity_id, Pointer{
+                    .archetype_index = ptr.archetype_index,
+                    .row_index = ptr.row_index,
+                });
+            }
 
-        // Either new storage (if the entity moved between storage tables due to having a new
-        // component) or the prior storage (if the entity already had the component and it's value
-        // is merely being updated.)
-        var current_archetype_storage = archetype_entry.value_ptr;
+            // Perform a swap removal to remove our entity from the archetype table.
+            archetype.remove(ptr.row_index);
 
-        if (new_hash == old_hash) {
-            // Update the value of the existing component of the entity.
+            _ = entities.entities.remove(entity);
+        }
+
+        /// Returns the archetype storage for the given entity.
+        pub inline fn archetypeByID(entities: *Self, entity: EntityID) *ArchetypeStorage {
             const ptr = entities.entities.get(entity).?;
-            current_archetype_storage.set(entities.allocator, ptr.row_index, name, component);
-            return;
+            return &entities.archetypes.values()[ptr.archetype_index];
         }
 
-        // Copy to all component values for our entity from the old archetype storage (archetype)
-        // to the new one (current_archetype_storage).
-        const new_row = try current_archetype_storage.appendUndefined(entities.allocator);
-        const old_ptr = entities.entities.get(entity).?;
-
-        // Update the storage/columns for all of the existing components on the entity.
-        current_archetype_storage.set(entities.allocator, new_row, "id", entity);
-        for (archetype.columns) |column| {
-            if (std.mem.eql(u8, column.name, "id")) continue;
-            for (current_archetype_storage.columns) |corresponding| {
-                if (std.mem.eql(u8, column.name, corresponding.name)) {
-                    const old_value_raw = archetype.getRaw(old_ptr.row_index, column.name);
-                    current_archetype_storage.setRaw(new_row, corresponding, old_value_raw) catch |err| {
-                        current_archetype_storage.undoAppend();
-                        return err;
-                    };
-                    break;
-                }
+        /// Sets the named component to the specified value for the given entity,
+        /// moving the entity from it's current archetype table to the new archetype
+        /// table if required.
+        pub fn setComponent(
+            entities: *Self,
+            entity: EntityID,
+            comptime namespace_name: anytype,
+            comptime component_name: anytype,
```

Comment:
```suggestion
            comptime namespace_name: @Type(.EnumLiteral),
            comptime component_name: @Type(.EnumLiteral),
```
Using a more constrained type here would better convey intent. You could also use `all_components` to derive the two types which constrains the interface further.